packages/kotti-ui/source/make-props.test.ts

```diff
@@ -1,51 +1,143 @@
 import { PropOptions } from '@vue/composition-api'
+import { castArray, isEqual } from 'lodash'
+import { PropType } from 'vue'
 import { z } from 'zod'
 
 import { makeProps } from './make-props'
 import { silenceConsole } from './test-utils/silence-console'
 
-const expectDefault = (
-	prop: PropOptions<unknown, boolean>,
-	defaultValue: unknown,
-) => {
-	// @ts-expect-error
-	expect(prop.default()).toEqual(defaultValue)
-	expect(prop.required).toBe(undefined)
-}
-
-const expectRequired = (prop: PropOptions<unknown, boolean>) => {
-	expect(prop.default).toEqual(undefined)
-	expect(prop.required).toBe(true)
-}
-
-const expectValidator = (
-	prop: PropOptions<unknown, boolean>,
-	successCases: unknown[],
-	failCases: unknown[],
-) => {
-	const { validator } = prop
-
-	if (validator === undefined) {
-		expect(validator).not.toBe(undefined)
-		throw new Error('unexpected undefined validator')
+declare global {
+	namespace jest {
+		interface Matchers<R> {
+			toBeRequired(): R
+			toBeType(expectedPropType: PropType<unknown>): R
+			toDefaultTo(defaultValue: unknown): R
+			toValidate(...cases: unknown[]): R
+		}
 	}
-
-	successCases.forEach((c) => expect(validator(c)).toBe(true))
-	failCases.forEach((c) => expect(validator(c)).toBe(false))
 }
 
+expect.extend({
+	toBeRequired(prop: PropOptions<unknown, boolean>) {
+		const passNoDefault = prop.default === undefined
+		const passRequired = prop.required === true
+
+		return {
+			message: () => {
+				if (!passNoDefault)
+					return `Expected “prop.default” to be “undefined”. Was “${prop.default}”`
+				if (!passRequired)
+					return `Expected “prop.required” to be “true”. Was “${prop.required}”`
+				return 'valid'
+			},
+			pass: passNoDefault && passRequired,
+		}
+	},
+	toBeType(prop: PropOptions<unknown, boolean>, expected: PropType<unknown>) {
+		const actualPropTypes = new Set(castArray(prop.type as PropType<unknown>))
+		const expectedPropTypes = new Set(castArray(expected))
+
+		const pass = isEqual(actualPropTypes, expectedPropTypes)
```

Comment:
I just tried 
```_.isEqual([1,2], [2,1])
```
and it returns `false`

which might be non-deterministic without an explicit `sort`

Reply:
It should be `false` as you’re comparing two non-equal arrays. Those are `Set`s

Reply:
It should be `false`
> wasn't saying it should be `true`

Those are `Set`s
> what are `Sets`? `castArray` returns an `Array`? 
> my point is more about how you construct the types outside of this function, to my knowledge, most Array prototype functions are "async" - thus, not respecting sorting, which might cause non-determinism

Reply:
Also, slight correction: Array prototype functions aren’t `async`, they’re entirely synchonous, but have a different scope than the outside.

E.g. `Array.prototype.forEach` is doing the same thing as `for (const item of array)`, besides that you can’t return/break/throw, etc.. That part is only the case because they use callbacks though, not because they’re actually async/running in the next V8 tick

Reply:
nvm, I didn't see `new Set(` might have been the diff I was on or just my idiocy.